### PR TITLE
fix: wait for storage sync before resolving the new account creation

### LINKED ISSUE #2161: ReactNative:  Error withLoadedAccount {"err": [Error: Account has no profile]}
I get 
```
 ERROR  Error withLoadedAccount {"err": [Error: Account has no profile]}
```
which seems to be from recent work on `cojson/src/localNode.ts` by you and @aeplay.  Any way you can take a look at why we don't have the "default profile" behavior like we used to?

_Originally posted by @boorad in https://github.com/garden-co/jazz/issues/2123#issuecomment-2860586004_

### PR DESCRIPTION
**Pull request by gdorsi**
Changes:
- Make waitForSync work on storage peers, adding content ack and switching to optimistic/known states
- Skip closed peers and not subscribed client peers when calling waitForSync
- Wait for storage sync before resolving the new account creation

This also fixes a critical bug on waitForSync by fixing `combineOptimisticWith` which should update only the optimistic state.

fixes #2161

### CODE REVIEW

File: packages/cojson/src/PeerState.ts [hunk 1]
```diff
@@ -76,53 +64,33 @@ export class PeerState {
...
+    this._optimisticKnownStates.combineWith(id, value);
   }
 
   combineOptimisticWith(id: RawCoID, value: CoValueKnownState) {
-    if (this._optimisticKnownStates === "assumeInfallible") {
-      this._knownStates.combineWith(id, value);
-    } else {
-      this._optimisticKnownStates.combineWith(id, value);
-    }
+    this._optimisticKnownStates.combineWith(id, value);
```
Comment (gdorsi) [inline packages/cojson/src/PeerState.ts:76]: This should update only the optimistic state.

File: packages/cojson/src/coValueCore/coValueCore.ts [hunk 2]
```diff
@@ -656,15 +652,15 @@ export class CoValueCore {
...
-    );
+    if (a.madeAt !== b.madeAt) {
+      return a.madeAt - b.madeAt;
+    }
+
+    if (a.txID.sessionID === b.txID.sessionID) {
+      return a.txID.txIndex - b.txID.txIndex;
+    }
+
+    return 0;
```
Comment (gdorsi) [inline packages/cojson/src/coValueCore/coValueCore.ts:663]: Fixes an edge case where two updates on two different sessions are happening at the same time.

After the latest optimizations was generating some false negatives on the tests.

Basically this way if the timestamp is the same but the sessions are not the insert order is preserved.

File: packages/cojson/src/coValueCore/coValueCore.ts [hunk 3]
```diff
@@ -1006,6 +1002,10 @@ export class CoValueCore {
       const waitingForPeer = new Promise<void>((resolve) => {
         const markNotFound = () => {
           if (this.peers.get(peer.id)?.type === "pending") {
+            logger.warn("Timeout waiting for peer to load coValue", {
+              id: this.id,
+              peerID: peer.id,
+            });
```
Comment (gdorsi) [inline packages/cojson/src/coValueCore/coValueCore.ts:1008]: Restored the warn on timeout.

File: packages/cojson/src/streamUtils.ts [hunk 4]
```diff
@@ -43,32 +37,11 @@ export function connectedPeers(
...
-          );
-          yield msg;
-        }
-      })(),
-      channel,
-    ];
-  } else {
-    return [channel.wrap(), channel];
-  }
+  return [channel.wrap(), channel];
```
Comment (gdorsi) [inline packages/cojson/src/streamUtils.ts:46]: Removed the debug code

File: packages/cojson/src/sync.ts [hunk 5]
```diff
@@ -622,28 +630,24 @@ export class SyncManager {
...
+      if (this.requestedSyncs.has(coValue.id)) {
+        this.syncCoValue(coValue);
+      }
+    });
 
-      this.requestedSyncs.set(coValue.id, promise);
-      return promise;
-    }
+    this.requestedSyncs.add(coValue.id);
   }
```
Comment (gdorsi) [inline packages/cojson/src/sync.ts:646]: Simplified to not use promises (weren't used)

Done to debug, and kept the refactoring because it simplifies the flow.

File: packages/cojson/src/sync.ts [hunk 6]
```diff
@@ -393,10 +399,12 @@ export class SyncManager {
...
-        // Should move the state to loading
-        this.local.loadCoValueCore(msg.id, peer.id).catch((e) => {
-          logger.error("Error loading coValue in handleLoad", { err: e });
-        });
+        // Syncronously updates the state loading is possible
+        coValue
+          .loadFromPeers(this.getServerAndStoragePeers(peer.id))
+          .catch((e) => {
+            logger.error("Error loading coValue in handleLoad", { err: e });
+          });
```
Comment (gdorsi) [inline packages/cojson/src/sync.ts:407]: Removed one layer of abstraction, helped on debugging and kept like this because it makes the flow easier to navigate

**Review by aeplay — APPROVED** [review]
